firebase-storage/src/main/java/com/google/firebase/storage/StorageReference.java

```diff
@@ -563,14 +566,109 @@ public StreamDownloadTask getStream(@NonNull StreamDownloadTask.StreamProcessor
+    TaskCompletionSource<ListResult> pendingResult = new TaskCompletionSource<>();
+    StorageTaskScheduler.getInstance()
+        .scheduleCommand(new ListTask(this, maxResults, /* pageToken= */ null, pendingResult));
+    return pendingResult.getTask();
+  }
+
+  /**
+   * Resumes a previous call to {@link #list(int)}, starting after a pagination token. Returns the
+   * next set of items (files) and prefixes (folders) under this StorageReference.
+   *
+   * <p>"/" is treated as a path delimiter. Firebase Storage does not support invalid object paths
+   * that end with "/" or contain two consecutive "//". All invalid objects in GCS will be filtered.
+   *
+   * @param maxResults The maximum number of results to return in a single page. Must be greater
+   *     than 0 and at most 1000.
+   * @param pageToken A page token from a previous call to list.
+   * @return A a {@link Task} that returns the next items and prefixes under the current
+   *     StorageReference.
+   */
+  @NonNull
+  @PublicApi
+  public Task<ListResult> list(int maxResults, @NonNull String pageToken) {
+    Preconditions.checkArgument(maxResults > 0, "maxResults must be greater than zero");
+    Preconditions.checkArgument(maxResults <= 1000, "maxResults must be less than 1000");
+    Preconditions.checkArgument(
+        pageToken != null, "pageToken must be non-null to resume a previous list() operation");
+    TaskCompletionSource<ListResult> pendingResult = new TaskCompletionSource<>();
+    StorageTaskScheduler.getInstance()
+        .scheduleCommand(new ListTask(this, maxResults, pageToken, pendingResult));
+    return pendingResult.getTask();
+  }
+
+  /**
+   * List all items (files) and prefixes (folders) under this StorageReference.
+   *
+   * <p>This is a helper method for calling list() repeatedly until there are no more results. The
+   * default pagination size is 1000. Consistency of the result is not guaranteed if objects are
+   * inserted or removed while this operation is executing.
+   *
+   * @throws OutOfMemoryError If there are too many items at this location.
+   * @return A {@link Task} that returns all items and prefixes under the current StorageReference.
+   */
+  @NonNull
+  @PublicApi
+  public Task<ListResult> listAll() {
+    TaskCompletionSource<ListResult> pendingResult = new TaskCompletionSource<>();
+
+    List<StorageReference> prefixes = new ArrayList<>();
+    List<StorageReference> items = new ArrayList<>();
+
+    Executor executor = StorageTaskScheduler.getInstance().getCommandPoolExecutor();
+    Task<ListResult> list = list(ListTask.DEFAULT_PAGE_SIZE);
+
+    Continuation<ListResult, Task<Void>> continuation =
+        new Continuation<ListResult, Task<Void>>() {
+          @Override
+          public Task<Void> then(@NonNull Task<ListResult> currentPage) {
+            ListResult result = currentPage.getResult();
+            prefixes.addAll(result.getPrefixes());
+            items.addAll(result.getItems());
```

Comment:
This is O(n2) for multiple pages.
`ListResult.Builder#appendListResult()` would be nice here.

Reply:
If we want to make this O(n) then we have to initialize the resulting array with the correct capacity upfront. I updated the code to do this, but am a little worried about the readability.

Reply:
I don't think correct capacity initialization is needed for amortized O(n).
Two growing lists should serve the purpose.

Reply:
There are two approaches to merging a lot of lists:

The original proposal with uses `addAll()` repeatedly. For ArrayList (LinkedList are better here, but I don't want to return a LinkedList-backed List to the consumer) the list [grows 2n every time](https://cs.corp.google.com/piper///depot/google3/third_party/java_src/jdk/openjdk8/u60/trunk/jdk/src/share/classes/java/util/ArrayList.java?rcl=249564389&g=0&l=257). If we are adding 9 pages of 1000 items:

- We start out with a list of 0 elements.
- Page 1: We increase the size to 1000 in one step. No existing elements are copied. 1000 new elements are copied.
- Page 2: We increase the size to 2000. The first 1000 elements are copied again.  1000 new elements are copied.
- Page 3: We increase the size to 4000. The first 2000 elements are copied again.  1000 new elements are copied.
- Page 4: 1000 new elements are copied.
- Page 5: We increase the size to 8000.  The first 4000 elements are copied again.  1000 new elements are copied.
- Page 6: 1000 new elements are copied.
- Page 7: 1000 new elements are copied.
- Page 8: 1000 new elements are copied.
- Page 9: We increase the size to 16000.  The first 8000 elements are copied again.  1000 new elements are copied.

Your suggestion of doing this inside the Builder seems equivalent to me (I might not be following correctly).

If we are setting the initial capacity, we reduce the number of copies to 9 * 1000 (down from 24 * 1000) and do not allocate space for 7000 unused elements. The amortized cost of both operations is O(n) (if I remember my interview training correctly), but O(n) might not be the best measurement for small datasets on Mobile devices that are much more resource constrained.

Reply:
The trade-off is (N is # of total result)
* List of List and merge in the end.
 * N copies + N creations
 * Max array N memory
* Two accumulating list
 * 2N~3N copies + N creations.
 * Max array 2N~4N memory
 * Better Readability

There is a constant performance improvement, but I would not sacrifice readability for it.
We could let the third-pair of eyes decide here.